Adding numpy-style broadcasting to mean operator

luisenp opened the pull request:
Summary: Adding numpy-style broadcast to mean operator

Differential Revision: D13766546

houseroad commented on caffe2/operators/mean_op.cc:
```diff
@@ -79,19 +79,21 @@ print("Mean:", workspace.FetchBlob("X"))
 
 OPERATOR_SCHEMA(MeanGradient)
-    .NumInputs(1)
+    .NumInputs(1, INT_MAX)
```
In the gradient op, you are propagating the gradients back. The input should be GO(0), and you only have one output for Mean op. So for MeanGradient, the input number should 1 as well.

luisenp replied:
The reason I'm adding extra inputs is to obtain the original sizes of the input tensors. I couldn't find another way to do this other than passing `I(0), I(1)`, etc as part of the input to the gradient operator (as far as I can tell, `GI(0)...GI(N)` seem to be all set by default to the fully broadcasted size). Note that this is similar to the [gradient of the elementwise_add operator](https://github.com/pytorch/pytorch/blob/d47108add0eaddbdaf1cf8bc95599faac66442cd/caffe2/operators/elementwise_add_gradient_op.cc#L24), which receives the two inputs even though the add operation only has one output. 

I'm happy to change this if there is a better way to do it, just not sure how that would be done.

BIT-silence replied:
That's a good point, it seems currently we don't have a better solution for the input sizes.

houseroad commented on caffe2/operators/mean_op.cc:
```diff
@@ -79,19 +79,21 @@ print("Mean:", workspace.FetchBlob("X"))
 class GetMeanGradient : public GradientMakerBase {
   using GradientMakerBase::GradientMakerBase;
   vector<OperatorDef> GetGradientDefs() override {
+    auto inputs = std::vector<string>();
```
I don't think you need to change GetMeanGradient.

luisenp replied:
See comment above.

houseroad commented on caffe2/operators/mean_op.h:
```diff
@@ -20,35 +21,45 @@ class MeanOp final : public Operator<Context> {
+                  output_dims.begin(),
+                  output_dims.begin() + output_dims.size(),
+                  input_dims[0].begin()),
+          "Cannot broadcast output to the first dimension.");
```
This error information does not match your check.

luisenp replied:
Would something like this be better? "Cannot broadcast output to the dimensions of the first input".

houseroad commented on caffe2/operators/mean_op.h:
```diff
@@ -72,6 +83,34 @@ class MeanOp final : public Operator<Context> {
   }
+
+ private:
+  std::vector<int64_t> ComputeBroadcastDims(
```
@BIT-silence do we have some existing code which can be reused in this case?

luisenp replied:
While @BIT-silence responds, one option I thought about was to use `elementwise_ops_util::ComputeBinaryBroadcastForwardDims` one dimension at a time (broadcasting Input i and the current Output obtained form Inputs 0 to i-1). 

Would that be preferable?

BIT-silence replied:
I think it's fine to make another function from the efficiency side. Since it will also be used for SumOp, I'm wondering if we can make it as another util function. But it's fine to just keep it here.

houseroad replied:
Ok, let's keep the current solution. Later we can merge them together.

houseroad commented on caffe2/operators/mean_op.h:
```diff
@@ -84,26 +123,39 @@ class MeanGradientOp : public Operator<Context> {
+        dY.sizes().cbegin(),
+        dY.sizes().cend(),
+        std::back_inserter(output_dims));
+    for (int i = num_inputs; i >= 1; i--) {
```
in case Mean(a, b, c), GI(1) (i.e., GI(b)) should be computed as GO(0) / input_num. If the shape doesn't match, (broadcasting happens), you should use reduce sum to get the final gradient for b. the same applies to a, c.

luisenp replied:
I think this is what the code is doing. Note that `dY = Input(0) = GO(0)`, and I use reduce sum in line 148 with input `dY` and the current input `dX = GI(i-1)` broadcasting it to dimensions specified by `Input(i) = I(i-1)` . One thing I'm not doing is checking if the dimensions are equal, to avoid the ReduceSum computation when it is unnecessary. 

Also note that the `op_mean_test.py` has a gradient check that includes a size check, and the test passes. @BIT-silence is there something missing on those tests?

houseroad reviewed (changes requested):
Thanks for the contribution. Several problems we need to address them before landing.

BIT-silence commented on caffe2/operators/mean_op.h:
```diff
@@ -84,26 +123,39 @@ class MeanGradientOp : public Operator<Context> {
+          dY_data,
+          dX_data,
+          &context_);
+      math::Scale(dX->numel(), scale, dX_data, dX_data, &context_);
```
There is no need to do this scale. Just pass scale to the 5th arg of math::ReduceSum.

BIT-silence commented on caffe2/operators/mean_op.h:
```diff
@@ -84,26 +123,39 @@ class MeanGradientOp : public Operator<Context> {
+          input_dims, output_dims, &input_axes, &output_axes);
+      auto* dX = Output(i - 1, original_input_size, at::dtype<T>());
+      auto* dX_data = dX->template mutable_data<T>();
+      math::ReduceSum(
```
```suggestion
      math::ReduceSum<T, Context>(
          output_dims.size(),
          output_dims.data(),
          input_axes.size(),
          input_axes.data(),
          scale,
          dY_data,
          dX_data,
          &context_);
```

BIT-silence commented on caffe2/operators/mean_op.h:
```diff
@@ -20,35 +21,45 @@ class MeanOp final : public Operator<Context> {
+          "Cannot broadcast output to the first dimension.");
+    }
+    auto* output_data = output->template mutable_data<T>();
+    std::vector<int> output_dims_int;
```
```suggestion
    const std::vector<int> output_dims_int(output_dims.cbegin(), output_dims.cend());
```

BIT-silence commented on caffe2/operators/mean_op.h:
```diff
@@ -20,35 +21,45 @@ class MeanOp final : public Operator<Context> {
+          (output_dims.size() == input_dims[0].size()) &&
+              std::equal(
+                  output_dims.begin(),
+                  output_dims.begin() + output_dims.size(),
```
```suggestion
                  output_dims.end(),
```

BIT-silence commented on caffe2/operators/mean_op.h:
```diff
@@ -20,35 +21,45 @@ class MeanOp final : public Operator<Context> {
-      }
+    std::vector<std::vector<int>> input_dims(InputSize());
+    for (int i = 0; i < InputSize(); i++) {
+      std::copy(
```
```suggestion
      input_dims[i].assign(Input(i).sizes.cbegin(), Input(i).sizes.cend());
```

BIT-silence commented on caffe2/operators/mean_op.h:
```diff
@@ -20,35 +21,45 @@ class MeanOp final : public Operator<Context> {
     auto& input0 = Input(0);
-
     auto* output = Output(0, input0.sizes(), at::dtype<T>());
     output->CopyFrom(input0, true /*async*/);
```
Here when InputSize == 2, that will not be very efficient since we have to copy and then add. Actually InputSize == 2 will be one of the most common usecases. So I'd suggest optimize it.

BIT-silence commented on caffe2/operators/mean_op.h:
```diff
@@ -84,26 +123,39 @@ class MeanGradientOp : public Operator<Context> {
-    int size = dY.numel();
-
     int num_inputs = OutputSize();
     float scale = 1.0f / num_inputs;
```
```suggestion
    cons T scale = T(1) / static_cast<T>(num_inputs);
```

BIT-silence commented on caffe2/operators/mean_op.h:
```diff
@@ -84,26 +123,39 @@ class MeanGradientOp : public Operator<Context> {
-      auto* cur_dX = Output(i);
-      cur_dX->ResizeLike(dY);
-      cur_dX->CopyFrom(*dX0, true /*async*/);
+    std::vector<int> output_dims;
```
```suggestion
    const std::vector<int> output_dims(dY.sizes().cbegin(), dY.sizes().cend());
```

luisenp commented:
@BIT-silence I tried to address all the suggested changes. Let me know if I missed anything.

houseroad commented on caffe2/operators/mean_op.h:
```diff
@@ -72,6 +114,34 @@ class MeanOp final : public Operator<Context> {
+        int isz = input_sizes[i];
+        if (isz >= 0) {
+          CAFFE_ENFORCE(
+              max_dim == -1 || max_dim == 1 || input_dims[i][isz] == max_dim ||
```
Actually, there is one weird behavior is if a's shape is [1, 1] and b's shape is [0], mean(a, b) should have shape [1, 0]. I know this is weird, but if we are following the numpy's broadcasting, we should do so. I am not sure how Add/Sub/... is implemented. I think we can fix this later. cc: @BIT-silence

houseroad reviewed (changes requested):
Confirmed with @BIT-silence, right now other elementwise ops follow numpy's style, [0] op [1] will have shape [0]. So let's fix this please, and also add one test case for this condition. Other part looks good to me.

luisenp committed:
Adding numpy-style broadcasting to mean operator (#16531)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/16531

Adding numpy-style broadcast to mean operator

Differential Revision: D13766546

fbshipit-source-id: 9dda556de4551e2f9f830bf57ed3a0f04bf2c35c
